request.c

```diff
@@ -0,0 +1,120 @@
+#include <stdlib.h>
+#include <arpa/inet.h>
+#include <stdint.h>
+#include <string.h>
+#include "request.h"
+
+DNSHeader* NewDNSHeader(u_int16_t id, u_int16_t flags, u_int16_t num_questions)
+{
+    DNSHeader *header     = calloc(1, sizeof(DNSHeader));
+    header->id            = htons(id);
+    header->flags         = htons(flags);
+    header->num_questions = htons(num_questions);
+
+    return header;
+};
+
+DNSQuestion* NewDNSQuestion(char *encoded_name, int type, int class)
+{
+    DNSQuestion *question  = calloc(1, sizeof(DNSQuestion));
+    question->encoded_name = malloc(strlen(encoded_name) + 1);
```

Comment:
optional nit: I recently learned that `malloc + strcpy` can be simplified to just `strdup`